andela/bp-esa-backend
# Create slack service to automate slack invite

#### What does this PR do?
It adds a part of the slack service that automates adding a user to a slack channel
#### Description of Task to be completed?
This Pull Request fixes issue #16 
#### How should this be manually tested?

- clone the repo
- `cd` into the project dir
- use the information in `.env.sample` to provide the necessary environment variables
- check the `setup.js` file in the Slack Service folder to tweak the channels to invite a user to
- use `npm run start:dev` to start the application
- note that the default port is `8000`
- send a `POST` request of type `application/x-www-form-urlencoded` with the email of a user(who is already in the test slack workspace) in the body of the request to `http://localhost:8000/api/v1/onboarding `

#### Any background context you want to provide?
- **Why underscore convention for variable naming**
Slack API uses the underscore convention for its variable naming. Everything from the body to be sent to different API endpoints, the response that would received from the API calls, etc all uses the underscore variable naming convention. It makes a lot of sense to also implement the Slack Service following that convention.
- **Why manually specifying the slack channels**
A very nice implementation of this feature would be to have the channel names specified as part of the body of the `POST` request to the onboarding endpoint. However, these are the reasons I believe that implementation is not good enough: 
- - Given that the names of slack channels can change at any time, I am not sure if we can rely on it for the functionality of this service.
- - It is a very costly/heavy operation to move from channel names to channel IDs which is what is really required in the implementation of this functionality. Presently, slack does not have an API endpoint that can be used to get a slack channel ID from its channel name, the only possible solution I have come across is to fetch all the channels in the slack workspace then run a `map` + `filter` on the returned collections to determine the ID of the channel we want.
- **Private channels alone**
 Presently, this feature can only be used to add slack members to private slack channels.
- **How to identify a channel ID**
You are required to provide the channel ID when specifying the channels you want the users to be automatically added to. Identifying the channel ID is as easy as identifying the channel name, this is a link to a helpful resource https://www.wikihow.com/Find-a-Channel-ID-on-Slack-on-PC-or-Mac
- **Required Slack API scopes:**
For the app to work as expected, enable the following scopes before installing the slack app:
- - users:read
- - users:read.email
- - group.write
- **Required Slack setup**
- - Ensure that you are in the channels you are trying to automate invite to because the app would be working on your behalf.
